pkg/resolvers/k8s/watcher.go

```diff
@@ -39,63 +54,100 @@ func startNewWatcher(
 	watcherErrs prometheus.Counter,
 	watcherGotChanges prometheus.Counter,
 ) (*watcher, error) {
-	// NOTE(bplotka): Would love to have proper context from above but naming.Resolver does not allow that.
 	ctx, cancel := context.WithCancel(context.Background())
-
-	s, err := startNewStreamer(ctx, target, epClient)
-	if err != nil {
-		return nil, err
-	}
-
 	return &watcher{
-		logger:     logger,
-		ctx:        ctx,
-		cancel:     cancel,
-		target:     target,
-		streamer:   s,
-		addrsState: map[string]struct{}{},
-
+		logger:            logger,
+		ctx:               ctx,
+		cancel:            cancel,
+		target:            target,
+		epClient:          epClient,
+		addrsState:        map[string]struct{}{},
 		resolvedAddrs:     resolvedAddrs,
 		watcherErrs:       watcherErrs,
 		watcherGotChanges: watcherGotChanges,
+		streamer:          nil,
+		streamRetryBackoff: &backoff.Backoff{
+			Min:    50 * time.Millisecond,
+			Jitter: true,
+			Factor: 2,
+			Max:    2 * time.Second,
+		},
 	}, nil
 }
 
 // Close closes the watcher, cleaning up any open connections.
 func (w *watcher) Close() {
+	if w.streamer != nil {
+		w.streamer.Close()
+	}
+
 	w.cancel()
 }
 
+func (w *watcher) startNewStreamerWithRetry(ctx context.Context) *streamer {
+	for ctx.Err() == nil {
```

Comment:
This should be something like:
```
for s, err := startNewStreamer(...); err != nil {
  s.close()
  w.logger...
  select {
    <- time.After(w.streamRetryBackoff.Duration()): 
    <- ctx.Done():
      return nil
  }
}
w.streamRetryBackoff.Reset()
return s
```
, so that we don't continue to sleep when context is cancelled.